### DISCUSSION TITLE
DB Management Schedule

### DISCUSSION THREAD

**Discussion by NicholasAntinori** [glowroot/glowroot#1092] [General]
Hi everyone I had a question, is it possible to create scheduled actions for DB compression and DB defragmentation of glowroot? 

thanks in advance

**Comment by Never-lab** [comment]
Following up on my own question from 2024 with what the code actually does today.

### Current behavior (embedded / H2)
- **Expired data deletion** is already scheduled (`ReaperRunnable`) based on Configuration → Storage retention.
- **Capped detail files** are compressed on write (LZF).
- **H2 Defrag / Compact** exist only as **manual** admin actions:
  - UI: Configuration → Storage → “Defrag H2 data” / “Compact H2 data”
  - API: `POST /backend/admin/defrag-h2-data` and `…/compact-h2-data`
  - Implementation runs H2 `SHUTDOWN COMPACT` and, while it runs, Glowroot does **not** capture new data (as the confirm dialog warns).

So: deleting expired rows does **not** shrink the `.mv.db` file by itself; reclaiming file space still needs an explicit compact/defrag. There is **no built-in schedule** for that today.

### Workaround today
- Tighten retention (less data → less need to compact). Related: Deployment profile presets (Dev/Prod) in https://github.com/glowroot/glowroot/pull/1188
- Or call the admin endpoints from an external cron during a quiet window.

### Possible enhancement
An **opt-in** scheduled compact/defrag (e.g. weekly off-hours), with clear “no capture during operation” semantics — similar to the manual buttons, just timed.

cc @trask @nowheresly — would a small opt-in scheduler for H2 compact/defrag be in scope for embedded, or do you prefer keeping this as a manual/ops responsibility (external cron against the existing admin APIs)?
